### PR TITLE
feat: (Gen AI) Update the `document_retrieval` sample for embeddings

### PR DESCRIPTION
**Pull request by Valeriy-Burlaka**
## Description

Updates the sample for the "_Get text embeddings for a snippet of text_" [section](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings#get_text_embeddings_for_a_snippet_of_text) of the embeddings [page](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings):

* Fix the `Example response`
* Provide more clear and idiomatic example of using the Vertex SDK
* Show alternative ways to provide the input for the `get_embeddings` method.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [x] **Tests** pass:  `nox -s py-3.8` && `nox -s py-3.11`
- [x] **Lint** pass:   `nox -s lint`
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: generative_ai/embeddings/document_retrieval_example.py [hunk 1]
```diff
@@ -19,27 +19,41 @@
 
 
 def embed_text() -> list[list[float]]:
-    """Embeds texts with a pre-trained, foundational model.
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:22]: This comment is misleading (we do not "embed" texts anywhere, we calculate vector embeddings for them).

File: generative_ai/embeddings/document_retrieval_example.py [hunk 2]
```diff
@@ -19,27 +19,41 @@
...
+    """Calculates embeddings for the given text inputs.
 
     Returns:
-        A list of lists containing the embedding vectors for each input text
+        A list of lists containing the embedding vectors for each input text.
     """
 
-    # A list of texts to be embedded.
-    texts = ["banana muffins? ", "banana bread? banana muffins?"]
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:29]: The motivation to shift to a longer, more complex texts is the following:

* The new texts are more representative of real-world documentation or help articles that might be used in a text embedding scenario. This should serve as a subtle hint at potential applications of Vertex AI in improving search and retrieval for various help documents.
* The longer, more detailed texts provide a better demonstration of how the embedding model can handle complex, multi-sentence inputs.
* The new text examples specifically mention Google products and services, to subtly showcase Google's ecosystem.

File: generative_ai/embeddings/document_retrieval_example.py [hunk 3]
```diff
@@ -19,27 +19,41 @@
...
+
     # The dimensionality of the output embeddings.
-    dimensionality = 256
-    # The task type for embedding. Check the available tasks in the model's documentation.
-    task = "RETRIEVAL_DOCUMENT"
+    dimensionality = 128
+    embeddings = model.get_embeddings(inputs, output_dimensionality=dimensionality)
 
-    model = TextEmbeddingModel.from_pretrained("text-embedding-004")
-    inputs = [TextEmbeddingInput(text, task) for text in texts]
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:36]: I don't think this was a good showcase of the `TextEmbeddingInput` class API. Also, the name of the "task" argument is `task_name`. While positional arguments work here, using the keyword arguments is more error-prone to future changes.

File: generative_ai/embeddings/document_retrieval_example.py [hunk 4]
```diff
@@ -19,27 +19,41 @@
...
-    kwargs = dict(output_dimensionality=dimensionality) if dimensionality else {}
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:37]: This line us utterly confusing. We create a `dict` with just one argument, only to expand it with `**` on the next line. The `dimensionality` is always defined, so there is no need for `else` 🤷

File: generative_ai/embeddings/document_retrieval_example.py [hunk 5]
```diff
@@ -19,27 +19,41 @@
...
-    embeddings = model.get_embeddings(inputs, **kwargs)
+    for embedding in embeddings:
+        print(embedding)
+        # Example response:
+        # TextEmbedding(values=[0.06830032169818878, 0.03944097459316254, ...], 'statistics': { ... }, ...)
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:55]: Each item in the resulting list is a `TextEmbedding` object

File: generative_ai/embeddings/document_retrieval_example.py [hunk 6]
```diff
@@ -19,27 +19,41 @@
...
+        TextEmbeddingInput(
+            text=(
+                "You'll find some apps on your Home screens, and all your apps in All Apps. "
+                "You can open apps, switch between apps, and find 2 apps at once."
+            ),
+            title="Find, open & close apps on a Pixel phone",
+            # The task type for embedding. Check the available tasks in the model's documentation.
+            task_type="RETRIEVAL_DOCUMENT"
+        ),
+        # Option2: Use plain text as embedding input
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:41]: @msampathkumar , I'd really like to showcase the 2nd option here, which is using a plain text string as an input.

However, when I read this sample as an "external viewer", I immediately have a question:  "What will be the `task_type` of this input if I use a plain string?". And, unfortunately,  I can't answer this question myself.
I searched through the Vertex SDK source code, and I didn't find any hints for this 🤷 My guess is that `RETRIEVAL_DOCUMENT` is the default type. How can I prove or disprove this? Thanks!

File: generative_ai/embeddings/document_retrieval_example.py [hunk 7]
```diff
@@ -19,27 +19,41 @@
...
+            "You can download no-charge and paid apps from Google Play on your Android phone or tablet. "
+            "We recommend that you get apps from Google Play, but you can also get them from other sources."
+        )
+    ]
+
     # The dimensionality of the output embeddings.
-    dimensionality = 256
-    # The task type for embedding. Check the available tasks in the model's documentation.
-    task = "RETRIEVAL_DOCUMENT"
+    dimensionality = 128
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:49]: Reduce dimensionality to be mindful of the end user's wallet. The response `metadata` shows `{'billableCharacterCount': 296.0}`, which is ~ $0.00006 (which should be fine)

File: generative_ai/embeddings/document_retrieval_example.py [hunk 8]
```diff
@@ -19,27 +19,41 @@
...
+    """Calculates embeddings for the given text inputs.
 
     Returns:
-        A list of lists containing the embedding vectors for each input text
+        A list of lists containing the embedding vectors for each input text.
     """
 
-    # A list of texts to be embedded.
-    texts = ["banana muffins? ", "banana bread? banana muffins?"]
+    model = TextEmbeddingModel.from_pretrained("text-embedding-004")
```
Comment (Valeriy-Burlaka) [inline generative_ai/embeddings/document_retrieval_example.py:28]: I'm not sure how/why this works without the standard call to `vertexai.init(...)` (but it works 🤷 ). Should I probably add this call to be future-proof?

File: generative_ai/embeddings/document_retrieval_example.py (hunk 6 above)
Comment (msampathkumar) [reply generative_ai/embeddings/document_retrieval_example.py:41]: `model.get_embeddings`'s doc string has that the input has to be a list of texts or `TextEmbeddingInput` objects to embed. 

Function definition also says `List[Union[str, TextEmbeddingInput]]`. So I believe, simple text input is fine.

Comment (msampathkumar) [inline generative_ai/embeddings/document_retrieval_example.py:41]:
```suggestion
        # Option2: Use plain text as input
```

File: generative_ai/embeddings/document_retrieval_example.py (hunk 4 above)
Comment (msampathkumar) [reply generative_ai/embeddings/document_retrieval_example.py:37]: I believe, the idea was to share info `kwargs` option.

Comment (msampathkumar) [reply generative_ai/embeddings/document_retrieval_example.py:37]: Unless it was not a hinderance or deprecated, keep it for now. At a later a time, we will re-review all code samples.

File: generative_ai/embeddings/document_retrieval_example.py (hunk 8 above)
Comment (msampathkumar) [reply generative_ai/embeddings/document_retrieval_example.py:28]: `vertexai.init` is a must for a first time setup. Doing it again is harmless but good to have option. So this is not a bug.

File: generative_ai/embeddings/document_retrieval_example.py (hunk 4 above)
Comment (Valeriy-Burlaka) [reply generative_ai/embeddings/document_retrieval_example.py:37]: > I believe, the idea was to share info kwargs option.

@msampathkumar , this comment is not clear to me. We're passing `**kwargs` in the line below, which is an equivalent of passing named arguments but in a more obscure way. I believe that this sample does not foster Python best practices and it's worth fixing it.

**Comment by Valeriy-Burlaka** [comment]
@msampathkumar, following the review comments to this PR, I can't decide if you deem these fixes helpful or not. If you believe that everything is OK with [the sample](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings#generative-ai-get-text-embedding-python_vertex_ai_sdk) in its current state, I can simply close this PR. Thanks!

**Comment by msampathkumar** [comment]
Hi @Valeriy-Burlaka  Thank you for your work on this. For now, marking this request as closed.

**Pull request closed without merging**